### ISSUE TITLE
Rails 7.0.3 changes `t.references` in 6.0 migrations from `t.integer` to `t.bigint` on PostgreSQL.

### ISSUE THREAD

**Issue by andrewculver** [rails/rails#45093]
After upgrading Rails from 7.0.2.4 to 7.0.3, `rake db:drop db:create db:migrate` is producing changes in our `db/schema.rb`, from our `6.0` migrations which I wouldn't expect after an upgrade. I can't tell whether this is a new bug being introduced or a previous bug being fixed.

### Steps to reproduce

 - Here is a newly generated example 7.0.2.4 application with some 6.0 migrations and 7.0 migrations: https://github.com/andrewculver/rails-migration-issue
 - Here is a PR on that same repository where we upgrade from 7.0.2.4 to 7.0.3 and ran `rake db:drop db:create db:migrate` to produce the delta in `db/schema.rb`: https://github.com/andrewculver/rails-migration-issue/pull/1/files

Here are the steps I performed to generate that repository and PR:

```bash
mkdir old-rails
cd old-rails
echo "source 'https://rubygems.org'" > "Gemfile"
echo "gem 'rails', '7.0.2.4'" >> "Gemfile"
bundle install
bundle exec rails -v
# Confirm it says "7.0.2.4"
rails new migration-issue-app --database=postgresql
cd migration-issue-app

rm ./Gemfile.lock
vim ./Gemfile
# Lock "rails" at "7.0.2.4".
bundle install

rails g model Project name:string
rails g model Idea project:references name:string
vim db/migrate/*_create_projects.rb
# Change migration version to "6.0".
vim db/migrate/*_create_ideas.rb
# Change migration version to "6.0".

rails g model Site name:string
rails g model Page site:references name:string

rake db:create
rake db:migrate

cat db/schema.rb
# Confirm that `project_id` is `t.integer` and `site_id` is `t.bigint`.

git add -A
git commit -m "Set up the test condition."

rm ./Gemfile.lock
vim ./Gemfile
# Lock "rails" at "7.0.3".
bundle install
bundle exec rails -v
# Confirm it says "7.0.3"

rake db:drop
rake db:create
rake db:migrate

git diff db/schema.rb
# `project_id` should not have changed, but it has gone from `t.integer` to `t.bigint`.
```

### Expected behavior
I would not expect the 6.0 migrations to start producing changes in `db/schema.rb` when rebuilding it from scratch.

### Actual behavior
I do see changes to `db/schema.rb`:

```diff
--- a/db/schema.rb
+++ b/db/schema.rb
@@ -15,7 +15,7 @@ ActiveRecord::Schema[7.0].define(version: 2022_05_14_034501) do
   enable_extension "plpgsql"
 
   create_table "ideas", force: :cascade do |t|
-    t.integer "project_id", null: false
+    t.bigint "project_id", null: false
     t.string "name"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
```

### System configuration
**Rails version**: `Rails 7.0.3`
**Ruby version**: `ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [arm64-darwin21]`

**Comment by fatkodima** [comment]
Thanks for the reproduction. `bigint` is the correct type for postgres and mysql since `5.1` migrations version. It was accidentally changed by https://github.com/rails/rails/pull/43295/files#diff-01b7232fa00908de002595608171bf83348e1898a24fb2af88d7d44175bacb63R163-R164 when trying to fix only for sqlite.

And this was fixed back in https://github.com/rails/rails/pull/42350. I think, it was left unnoticed because not many projects rerun older migrations.

**Comment by yahonda** [comment]
This is expected behavior since https://github.com/rails/rails/pull/42350 has been backported to 7-0-stable branch via https://github.com/rails/rails/commit/1fe808f5d217726fcb738375fe6aa6ee407d8058#diff-fac852e063eee718c2af44dec6f80b1b46f52f8aae7936e5eea6b2eab08cf2f7 and has been released as Rails 7.0.3.

```ruby
% git diff v7.0.2.4 v7.0.3 activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb b/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
index 2cbe264fa6..1ea93c87d1 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb
@@ -202,6 +202,10 @@ def polymorphic_index_name(table_name)

         def index_options(table_name)
           index_options = as_options(index)
+
+          # legacy reference index names are used on versions 6.0 and earlier
+          return index_options if options[:_uses_legacy_reference_index_name]
+
           index_options[:name] ||= polymorphic_index_name(table_name) if polymorphic
           index_options
         end
%
```

**Comment by yahonda** [comment]
Closing this issue. Thanks for the detailed report.
